java/ql/src/Telemetry/APIUsage.qll

```diff
@@ -0,0 +1,44 @@
+import java
+private import semmle.code.java.dataflow.FlowSteps
+private import semmle.code.java.dataflow.ExternalFlow
+
+string jarName(CompilationUnit cu) {
+  result = cu.getParentContainer().toString().regexpCapture(".*/(.*\\.jar)/?.*", 1)
+}
+
+predicate isJavaRuntime(Callable call) {
+  jarName(call.getCompilationUnit()) = "rt.jar" or
+  call.getCompilationUnit().getParentContainer().toString().substring(0, 14) = "/modules/java."
+}
+
+// TODO Is this heuristic too broad?
+predicate isInterestingAPI(Callable call) {
+  call.getNumberOfParameters() > 0 and
+  not (
+    call.getReturnType() instanceof VoidType or
+    call.getReturnType() instanceof PrimitiveType or
+    call.getReturnType() instanceof BoxedType
+  )
+}
+
+// TODO [bm] Fails to detect Collection flow yet (e.g. Map#put)
+string supportKind(Callable api) {
+  if api instanceof TaintPreservingCallable
+  then result = "taint-preserving"
+  else
+    if
+      summaryModel(api.getCompilationUnit().getPackage().toString(),
+        api.getDeclaringType().toString(), _, api.getName(), _, _, _, _, _)
+    then result = "summary"
+    else
+      if
+        sinkModel(api.getCompilationUnit().getPackage().toString(),
+          api.getDeclaringType().toString(), _, api.getName(), _, _, _, _)
+      then result = "sink"
+      else
+        if
+          sourceModel(api.getCompilationUnit().getPackage().toString(),
+            api.getDeclaringType().toString(), _, api.getName(), _, _, _, _)
+        then result = "source"
+        else result = "?"
```

Comment:
These aren't precisely good places to hook into the existing code in order to draw conclusions about what is being modelled, but there isn't a completely clear-cut alternative. However, I think it would be much better to use something like
```ql
  api instanceof SummarizedCallable
  or
  exists(Call call, DataFlow::Node arg |
    call.getCallee() = api and
    [call.getAnArgument(), call.getQualifier()] = arg and
    TaintTracking::localAdditionalTaintStep(arg, _)
  )
```
for the summaries, `sinkNode(node, _)` for the sinks, and `RemoteFlowSource` for the sources.

Reply:
I've tried your suggestion in c48586ff8035a1ba803036537f0c20eb28fb6991 but faced a drop of results for various projects. I've reverted that change for now in 26d426907116636039df76eb1b2620e843a0fb51 but would love to have a look at this together once you have time.